src/Types/RelationDynamicMethodReturnTypeExtension.php

```diff
@@ -61,38 +67,65 @@ public function getTypeFromMethodCall(
+        $templateTypes = [];
 
-        if (! $argType instanceof ConstantStringType) {
+        // Determine TRelatedModel; this is the 1st parameter, if given, or `Eloquent\Model`
+        $relatedModelClassArgType = $numArgs === 0 ?
+            new ConstantStringType(Model::class, true) :
+            $scope->getType($methodCall->getArgs()[0]->value);
+        if (! $relatedModelClassArgType instanceof ConstantStringType) {
             return $returnType;
         }
+        $relatedModelClassName = $relatedModelClassArgType->getValue();
+        if (! $this->provider->hasClass($relatedModelClassName)) {
+            $relatedModelClassName = Model::class;
+        }
+        $templateTypes[] = new ObjectType($relatedModelClassName);
 
-        $argClassName = $argType->getValue();
-
-        if (! $this->provider->hasClass($argClassName)) {
-            $argClassName = Model::class;
+        // Determine TDeclaringModel; this is the model on whose instance the method is called
+        $calledOnType = $scope->getType($methodCall->var);
+        if (! $calledOnType instanceof TypeWithClassName) {
+            return $returnType;
+        }
+        $declaringModelClassName = $calledOnType->getClassName();
+        $templateTypes[] = new ObjectType($declaringModelClassName);
+
+        // Determine TIntermediateModel for "Through" types; this is the 2nd parameter
+        $hasManyThroughType = new ObjectType(HasManyThrough::class);
+        if ($hasManyThroughType->isSuperTypeOf($returnType)->yes()) {
+            $intermediateModelClassArgType = $scope->getType($methodCall->getArgs()[1]->value);
+            if (! $intermediateModelClassArgType instanceof ConstantStringType) {
+                return $returnType;
+            }
+            $intermediateModelClassName = $intermediateModelClassArgType->getValue();
+            $templateTypes[] = new ObjectType($intermediateModelClassName);
         }
 
-        // Special case for BelongsTo. We need to add the child model as a generic type also.
-        if ((new ObjectType(BelongsTo::class))->isSuperTypeOf($returnType)->yes()) {
-            return new GenericObjectType($returnType->getClassName(), [new ObjectType($argClassName), $calledOnType]);
+        // Work-around for a Laravel bug
+        // Opposed to other `HasOne...` and `HasMany...` methods,
+        // `HasOneThrough` and `HasManyThrough` do not extend a common
+        // `HasOneOrManyThrough` base class, but `HasOneThrough` directly
+        // extends `HasManyThrough`.
+        // This does not only violate Liskov's Substitution Principle but also
+        // has the unfortunate side effect that `HasManyThrough` cannot
+        // bind the template parameter `TResult` to a Collection, but needs
+        // to keep it unbound for `HasOneThrough` to overwrite it.
+        // Hence, if `HasManyTrough` is used directly, we must bind the
+        // fourth template parameter `TResult` here.
```

Comment:
@nagmat84 I encourage you to send a PR to Laravel fixing this problem.

Reply:
I will do so. That was my initial idea. However, I cannot anticipate how long it will take for Laravel to fix that bug, so I submitted this PR nonetheless. As soon as the said bug is fixed by Laravel, I will post a new PR here, which removes the then unnecessary code.